### PR TITLE
=wip Convert UpgardeStage to GraphStage

### PR DESCRIPTION
**Pull request by pchordia15**
#19361

### CODE REVIEW

**Comment by ktoso** [comment]
OK TO TEST

**Comment by pchordia15** [comment]
@ktoso can you please help me with what the error is? TIA

**Comment by ktoso** [comment]
It does not compile:

```
[error] /localhome/jenkinsakka/workspace/pr-validator-per-commit-jenkins/akka-http-core/src/main/scala/akka/http/impl/engine/ws/WebSocketClientBlueprint.scala:101: not found: value become
[error]                       become(transparent)
[error]                       ^
[error] one error found
```

You can find the errors by clicking here:

_wip_convert_upgardestage_to_graphstage_by_pchordia15_ _pull_request__20536_ _akka_akka

and then here:
pr-validator-per-commit-jenkins__6213_console__jenkins_

then search for `^` or `***` to find compile and test failures :) Hope this helps.

**Comment by pchordia15** [comment]
Thank you for helping me out with finding the error. I am not sure how to resolve the error for become as I am not getting what I should replace it with.

File: akka-http-core/src/main/scala/akka/http/impl/engine/ws/WebSocketClientBlueprint.scala [hunk 1]
```diff
@@ -57,68 +57,82 @@ object WebSocketClientBlueprint {
...
+
+      val out: Outlet[ByteString] = Outlet("UpgradeStage.out")
+      val in: Inlet[ByteString] = Inlet("UpgradeStage.in")
+
+      override val shape: FlowShape[ByteString, ByteString] = FlowShape(in, out)
+
+      override def createLogic(inheritedAttributes: Attributes): GraphStageLogic =
+        new GraphStageLogic(shape) with InHandler with OutHandler {
+
+          def initial: InHandler = parsingResponse
```
Comment (ktoso) [inline akka-http-core/src/main/scala/akka/http/impl/engine/ws/WebSocketClientBlueprint.scala]: initial does nothing in `GraphStage`, you have to set the in handler, like this instead: `setHandler(in, parsingStage)`

File: akka-http-core/src/main/scala/akka/http/impl/engine/ws/WebSocketClientBlueprint.scala [hunk 2]
```diff
@@ -57,68 +57,82 @@ object WebSocketClientBlueprint {
...
+              val elem = grab(in)
+              parser.parseBytes(elem) match {
+                case NeedMoreData ⇒ pull(in)
+                case ResponseStart(status, protocol, headers, entity, close) ⇒
+                  val response = HttpResponse(status, headers, protocol = protocol)
+                  Handshake.Client.validateResponse(response, subprotocol.toList, key) match {
+                    case Right(NegotiatedWebSocketSettings(protocol)) ⇒
+                      result.success(ValidUpgrade(response, protocol))
+
+                      become(transparent)
```
Comment (ktoso) [inline akka-http-core/src/main/scala/akka/http/impl/engine/ws/WebSocketClientBlueprint.scala]: as you now know, become is gone (please compile the code before submitting PRs :-)), so we'll need to replace it with `setHandler(in, transparent)` :)

File: akka-http-core/src/main/scala/akka/http/impl/engine/ws/WebSocketClientBlueprint.scala [hunk 3]
```diff
@@ -57,68 +57,82 @@ object WebSocketClientBlueprint {
...
-      }
 
-      def transparent: State = new State {
-        def onPush(elem: ByteString, ctx: Context[ByteString]): SyncDirective = ctx.push(elem)
-      }
+          def transparent: InHandler = new InHandler {
+            val elem = grab(in)
+            def onPush(): Unit = push(out, elem)
+          }
+          setHandlers(in, out, this)
```
Comment (ktoso) [inline akka-http-core/src/main/scala/akka/http/impl/engine/ws/WebSocketClientBlueprint.scala]: hm not sure if that's valid. this should mimic what `initial` was doing. Do we have the same logic invoked `initially`?

**Comment by ktoso** [comment]
Ping after a week, are you planning to continue with this PR? :)

**Comment by danxmoran** [comment]
@pchordia15 @ktoso I can pick this up, if this PR is dead.

**Comment by ktoso** [comment]
@danxmoran yeah, it would seem we've given enough time to finish it up, I'm going to close this PR - sorry @pchordia15 we weren't able to get it over the finish line.

Please have a shot at it @danxmoran, thanks a lot for the contrib in advance!

**Pull request closed without merging**